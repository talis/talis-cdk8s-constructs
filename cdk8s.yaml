language: typescript
imports:
  - k8s@1.30.0
