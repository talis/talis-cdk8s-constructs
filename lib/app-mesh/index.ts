export * from "./ingress-gateway-nlb";
export * from "./app-service";
