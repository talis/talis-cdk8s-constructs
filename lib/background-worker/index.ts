export * from "./background-worker-props";
export * from "./background-worker";
