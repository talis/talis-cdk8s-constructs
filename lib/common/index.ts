export * from "./container-props";
export * from "./env-util";
export * from "./name-util";
export * from "./statefulset-util";
export * from "./value-util";
export * from "./workload-props";
