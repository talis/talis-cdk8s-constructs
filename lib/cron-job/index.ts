export * from "./cron-job-props";
export * from "./cron-job";
