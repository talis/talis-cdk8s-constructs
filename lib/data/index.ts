export * from "./config-map";
export * from "./secret";
export * from "./factory";
