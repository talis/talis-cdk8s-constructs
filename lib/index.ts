export * from "./cron-job";
export * from "./background-worker";
export * from "./data";
export * from "./job";
export * from "./redis";
export * from "./talis-chart";
export * from "./web-service";
