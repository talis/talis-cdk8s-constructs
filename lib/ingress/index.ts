export * from "./ingress-props";
export * from "./ingress";
