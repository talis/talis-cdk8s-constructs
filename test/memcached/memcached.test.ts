import { Chart, Testing } from "cdk8s";
import { KubeService, KubeStatefulSet } from "../../imports/k8s";
import { Memcached, MemcachedProps } from "../../lib";
import { makeChart } from "../test-util";

const requiredProps = {
  release: "v1",
};

function synthMemcached(props: MemcachedProps = requiredProps) {
  const chart = Testing.chart();
  new Memcached(chart, "memcached-test", props);
  const results = Testing.synth(chart);
  return results;
}

describe("Memcached", () => {
  describe("Props", () => {
    test("Minimal required props", () => {
      const chart = Testing.chart();
      new Memcached(chart, "memcached-test", requiredProps);
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();
    });

    test("All the props", () => {
      const app = Testing.app();
      const chart = new Chart(app, "test", {
        namespace: "test",
        labels: {
          app: "my-app",
          environment: "test",
          region: "local",
        },
      });
      const selectorLabels = {
        app: "my-app",
        role: "memcached",
        instance: "test",
      };

      const allProps: Required<MemcachedProps> = {
        ...requiredProps,
        selectorLabels,
      };
      new Memcached(chart, "memcached-test", allProps);
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();

      const statefulSet = results.find((obj) => obj.kind === "StatefulSet");
      expect(statefulSet).toHaveAllProperties(allProps, [
        "release",
        "selectorLabels",
      ]);
    });

    test("selectorLabels can override app", () => {
      const results = synthMemcached({
        ...requiredProps,
        selectorLabels: { app: "foobar" },
      });
      const sts = results.find((obj) => obj.kind === "StatefulSet");
      expect(sts).toHaveProperty("metadata.labels.app", "foobar");
      expect(sts).toHaveProperty("spec.selector.matchLabels.app", "foobar");
      expect(sts).toHaveProperty("spec.template.metadata.labels.app", "foobar");
    });

    test("selectorLabels can override role", () => {
      const results = synthMemcached({
        ...requiredProps,
        selectorLabels: { role: "cache" },
      });
      const sts = results.find((obj) => obj.kind === "StatefulSet");
      expect(sts).toHaveProperty("metadata.labels.role", "cache");
      expect(sts).toHaveProperty("spec.selector.matchLabels.role", "cache");
      expect(sts).toHaveProperty("spec.template.metadata.labels.role", "cache");
    });
  });

  describe("Object instances", () => {
    test("Exposes service object through property", () => {
      const chart = makeChart();
      const memcached = new Memcached(chart, "memcached-test", requiredProps);
      expect(memcached.service).toBeDefined();
      expect(memcached.service).toBeInstanceOf(KubeService);
      expect(memcached.service.name).toEqual("memcached-test");
    });

    test("Exposes statefulSet object through property", () => {
      const chart = makeChart();
      const memcached = new Memcached(chart, "memcached-test", requiredProps);
      expect(memcached.statefulSet).toBeDefined();
      expect(memcached.statefulSet).toBeInstanceOf(KubeStatefulSet);
      expect(memcached.statefulSet.name).toEqual("memcached-test-sts");
    });
  });

  describe("Container release", () => {
    test("Default container release", () => {
      const results = synthMemcached();
      const sts = results.find((obj) => obj.kind === "StatefulSet");
      expect(sts).toHaveProperty(
        "spec.template.spec.containers[0].image",
        "memcached:v1"
      );
    });

    test("Container release set explicitly", () => {
      const results = synthMemcached({
        ...requiredProps,
        release: "12345",
      });
      const sts = results.find((obj) => obj.kind === "StatefulSet");
      expect(sts).toHaveProperty(
        "spec.template.spec.containers[0].image",
        "memcached:12345"
      );
    });
  });

  describe("getDnsName", () => {
    test("Builds a DNS name for the first Pod", () => {
      const chart = makeChart();
      const memcached = new Memcached(chart, "memcached-test", requiredProps);
      expect(memcached.getDnsName()).toBe(
        "memcached-test-sts-0.memcached-test"
      );
    });

    const tests: [number, string][] = [
      [0, "memcached-test-sts-0.memcached-test"],
      [1, "memcached-test-sts-1.memcached-test"],
      [3, "memcached-test-sts-3.memcached-test"],
    ];
    tests.forEach(([replica, expected]) => {
      test("Builds a string from non-empty parts", () => {
        const chart = makeChart();
        const memcached = new Memcached(chart, "memcached-test", requiredProps);
        expect(memcached.getDnsName(replica)).toBe(expected);
      });
    });
  });
});
