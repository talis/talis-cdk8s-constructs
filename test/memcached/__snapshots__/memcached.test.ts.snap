// Vitest Snapshot v1

exports[`Memcached > Props > All the props 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "labels": {
        "app": "my-app",
        "environment": "test",
        "instance": "test",
        "region": "local",
        "release": "v1",
        "role": "memcached",
      },
      "name": "test-memcached-test-c803ac31",
      "namespace": "test",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 11211,
          "protocol": "TCP",
        },
      ],
      "selector": {
        "app": "my-app",
        "instance": "test",
        "role": "memcached",
      },
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": {
      "labels": {
        "app": "my-app",
        "environment": "test",
        "instance": "test",
        "region": "local",
        "release": "v1",
        "role": "memcached",
      },
      "name": "test-memcached-test-memcached-test-sts-c8eb6f20",
      "namespace": "test",
    },
    "spec": {
      "replicas": 1,
      "selector": {
        "matchLabels": {
          "app": "my-app",
          "instance": "test",
          "role": "memcached",
        },
      },
      "serviceName": "test-memcached-test-c803ac31",
      "template": {
        "metadata": {
          "labels": {
            "app": "my-app",
            "environment": "test",
            "instance": "test",
            "region": "local",
            "release": "v1",
            "role": "memcached",
          },
        },
        "spec": {
          "containers": [
            {
              "args": [
                "--memory-limit=64",
              ],
              "image": "memcached:v1",
              "livenessProbe": {
                "failureThreshold": 5,
                "initialDelaySeconds": 5,
                "tcpSocket": {
                  "port": 11211,
                },
                "timeoutSeconds": 5,
              },
              "name": "memcached",
              "ports": [
                {
                  "containerPort": 11211,
                },
              ],
              "readinessProbe": {
                "failureThreshold": 5,
                "initialDelaySeconds": 5,
                "tcpSocket": {
                  "port": 11211,
                },
                "timeoutSeconds": 5,
              },
              "resources": {
                "limits": {
                  "cpu": "50m",
                  "memory": "64Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
]
`;

exports[`Memcached > Props > Minimal required props 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "labels": {
        "instance": "memcached-test",
        "release": "v1",
        "role": "memcached",
      },
      "name": "test-memcached-test-c803ac31",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 11211,
          "protocol": "TCP",
        },
      ],
      "selector": {
        "instance": "memcached-test",
        "role": "memcached",
      },
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": {
      "labels": {
        "instance": "memcached-test",
        "release": "v1",
        "role": "memcached",
      },
      "name": "test-memcached-test-memcached-test-sts-c8eb6f20",
    },
    "spec": {
      "replicas": 1,
      "selector": {
        "matchLabels": {
          "instance": "memcached-test",
          "role": "memcached",
        },
      },
      "serviceName": "test-memcached-test-c803ac31",
      "template": {
        "metadata": {
          "labels": {
            "instance": "memcached-test",
            "release": "v1",
            "role": "memcached",
          },
        },
        "spec": {
          "containers": [
            {
              "args": [
                "--memory-limit=64",
              ],
              "image": "memcached:v1",
              "livenessProbe": {
                "failureThreshold": 5,
                "initialDelaySeconds": 5,
                "tcpSocket": {
                  "port": 11211,
                },
                "timeoutSeconds": 5,
              },
              "name": "memcached",
              "ports": [
                {
                  "containerPort": 11211,
                },
              ],
              "readinessProbe": {
                "failureThreshold": 5,
                "initialDelaySeconds": 5,
                "tcpSocket": {
                  "port": 11211,
                },
                "timeoutSeconds": 5,
              },
              "resources": {
                "limits": {
                  "cpu": "50m",
                  "memory": "64Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
]
`;
